Add Mat3, Vec3, and some Quaternion tests

nascheinkman opened the pull request:
#49 I ported over some tests from the original cannon repository and added some more to try to give each method their own test and increase code coverage. 

While writing the tests, a few things came up:

1. Mat3.getTrace was returning void, but the doc said it should return Vec3. I made the function match the doc in commit 92083c5
2. A lot of methods allowed optional "target" objects to store the response in. Most methods did this by creating an argument with a default value of an empty object. Some functions seemed to recreate that using overloaded function definitions. I didn't think that was intentional so I made default value of empty objects consistent in commit 83050fa. 
3. Mat3.mmult and Mat3.transpose were failing the tests. If you try to multiply or transpose a matrix with itself as the argument, the data gets changed while the calculations happen creating wrong results. Commit 6ca03fb changes them to pass the tests, although this adds a little performance overhead with a Mat3.copy call each time. 

I'm hoping to write more tests for this project in the future. Is there a way that you want me to organize that? I was planning on creating pull requests as I go.

Thanks for your time.

marcofugaro commented:
Thanks for making this PR! This is great! I like that you kept the order of the methods from the source files.

One thing you should do however, is remove the tests with `//Test for changing target vector`, where you pass the second optional parameters to the method. Both for the Vec3 and Mat3. This is a pattern that might be removed from cannon-es in the future. For now the tests with only one argument are enough.

> Mat3.getTrace was returning void, but the doc said it should return Vec3. I made the function match the doc in commit 92083c5

Alright, this is fine.

> A lot of methods allowed optional "target" objects to store the response in. Most methods did this by creating an argument with a default value of an empty object. Some functions seemed to recreate that using overloaded function definitions. I didn't think that was intentional so I made default value of empty objects consistent in commit 83050fa.

Could you leave this out of this PR? This belongs to a bigger subject. Your suggestions make total sense though.

> Mat3.mmult and Mat3.transpose were failing the tests. If you try to multiply or transpose a matrix with itself as the argument, the data gets changed while the calculations happen creating wrong results. 

Could you go into more details with this one? The cannon.js tests were failing, or the new ones?
Why would you want to multiply a matrix by itself? And why would you want to do `mat.transpose(mat)` instead of `mat.transpose()`?

nascheinkman commented:
Thanks for getting back to me quick! I just reverted the change in function definitions, and I can get started on removing the tests for passing target vectors. The old tests were not failing, just new ones I wrote for that specific case, so if the target pattern does go it would pass the tests. 

> This is a pattern that might be removed from cannon-es in the future.

I would prefer if the target pattern did stay because I think that it increases performance in most situations. I'm not certain and haven't run any benchmarks, but I'm pretty sure that passing in a target object for the result removes the need for javascript to allocate memory for a new object each time the function is called. I see that pattern used in cannon-es a lot, such as with Spring.applyForce allocating all of its Vec3s to private variables in the module to avoid creating them each time applyForce is called. I've been using that pattern in cannon-es in my own project and it seems to increase performance by a fair amount for code that's run every physics step. 

> Why would you want to multiply a matrix by itself?

Multiplying a matrix by itself is useful in some contexts for a user of the library, such as for an adjacency matrix. It's not explicitly called out as not being allowed in the doc, so the current behavior would be unexpected. 

> And why would you want to do `mat.transpose(mat)` instead of `mat.transpose()`?

If I was using the transpose for an intermediate calculation, and didn't care about keeping the original matrix, I could either do `tmp = tmp.transpose()` or `tmp.transpose(tmp)`. If I did `tmp = tmp.transpose()`, I believe javascript would create a new Mat3, calculate the transpose, change the reference variable `tmp` to the new Mat3, and leave the old Mat3 to garbage collection. Whereas if I did `tmp.transpose(tmp)` it would only calculate the transpose to change the Mat3 in place.

marcofugaro commented:
> I would prefer if the target pattern did stay because I think that it increases performance in most situations. 

Oops, I did not meant removed :) I meant it will be adapted to the three.js [Vector3](https://threejs.org/docs/index.html#api/en/math/Vector3) and [Matrix3](https://threejs.org/docs/index.html#api/en/math/Matrix3) behaviour. They also have a lot of optimizations of that kind in place.

> Multiplying a matrix by itself is useful in some contexts for a user of the library, such as for an adjacency matrix. It's not explicitly called out as not being allowed in the doc, so the current behavior would be unexpected.

On that note, how is three.js' [Matrix3.multiply](https://threejs.org/docs/index.html#api/en/math/Matrix3.multiply) handling this?

> If I was using the transpose for an intermediate calculation, and didn't care about keeping the original matrix, I could either do tmp = tmp.transpose() or tmp.transpose(tmp). If I did tmp = tmp.transpose(), I believe javascript would create a new Mat3, calculate the transpose, change the reference variable tmp to the new Mat3, and leave the old Mat3 to garbage collection. Whereas if I did tmp.transpose(tmp) it would only calculate the transpose to change the Mat3 in place.

The correct behaviour would be to always modify the matrix in place, like three.js. Matrix operations are expensive and it makes sense to always default on non-creating a new matrix. The user can create a new matrix if he wants with `.clone()`. What do you say?

nascheinkman commented:
> Oops, I did not mean removed :) I meant it will be adapted to the three.js Vector3 and Matrix3 behaviour. They also have a lot of optimizations of that kind in place.

That's good to hear! I removed the `//Test for changing target vector` like you asked in commit 3f81276. 
> On that note, how is three.js' Matrix3.multiply handling this?

I looked at their source code, and they copy the nine numbers of each matrix to local constants (a total of 18 constants labeled a11-a33 and b11-b33), and then set each element of the resultant matrix manually through multiplying and adding those constants in nine different lines. Their transpose function is doing something similar of copying all the primitives to a local variable and then modifying in place. 

In commit ad1f278, I changed the implementation of Mat3.transpose and Mat3.mmult to something similar to avoid the `.copy` function call while still passing the `mmult in place` and `transpose in place` tests.
> The correct behaviour would be to always modify the matrix in place, like three.js. Matrix operations are expensive and it makes sense to always default on non-creating a new matrix. The user can create a new matrix if he wants with `.clone()`. What do you say?

I like that! That sounds faster and more intuitive.

marcofugaro reviewed (approved):
Looks good now! Thanks!
